README.md

**题记**
  ——  执剑天涯，从你的点滴积累开始，所及之处，必精益求精，即是折腾每一天。
  
**重要消息**

* [网易云【玩转大前端】配套课程](https://study.163.com/instructor/1021406098.htm)
* [EDU配套  教程](https://edu.csdn.net/lecturer/1555)

* [Flutter开发的点滴积累系列文章](https://blog.csdn.net/zl18603543572/article/details/93532582)

***

### 1 使用封装库
#### 1.1 添加依赖
小编已将本效果写成一个依赖库，读者可根据使用直接添加依赖，代码如下：
github方式依赖

```java
  flutter_search_bars:
    git:
      url: https://github.com/zhaolongs/flutter_search_bar.git
      ref: master
```
pub仓库依赖 [查看最新版本](https://pub.flutter-io.cn/packages/flutter_search_bars)

```java
  #搜索框
  flutter_search_bars: ^0.0.1
```

#### 1.2 导包
在使用到搜索框 SearchBar 的地方导包如下：

```java
import 'package:flutter_search_bars/flutter_search_bars.dart';
```

#### 1.3 使用

##### 1.3.1 默认大小使用

```java
///这里只是用于显示的搜索框不用做输入
///参数[heroTag]用于页面过渡动画tag
///参数clickCallBack为当前搜索框点击事件回调
SearchStaticBar(
	heroTag: "searchStatidBar",
	clickCallBack: () {
	  NavigatorUtils.pushPage(context, TestPage2());
	},
)
```

##### 1.3.2 指定大小使用

```java
 Container(
     width: 120,
      ///这里只是用于显示的搜索框不用做输入
      ///参数[heroTag]用于页面过渡动画tag
      ///参数clickCallBack为当前搜索框点击事件回调
      child: SearchStaticBar(
        heroTag: "searchStatidBar",
        clickCallBack: () {
          NavigatorUtils.pushPage(context, TestPage2());
        },
      ),
    )
```
运行效果如下：
![在这里插入图片描述](https://img-blog.csdnimg.cn/20200625151652610.png?x-oss-process=image/watermark,type_ZmFuZ3poZW5naGVpdGk,shadow_10,text_aHR0cHM6Ly9ibG9nLmNzZG4ubmV0L3psMTg2MDM1NDM1NzI=,size_16,color_FFFFFF,t_70)

CHANGELOG.md
## [0.0.3] 左侧返回键

## [0.0.2] 补充文档 

## [0.0.1] - TODO: Add release date.

* TODO: Describe initial release.


lib/src/search_textfield_bar.dart
import 'dart:ui';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

/**
 * 创建人： Created by zhaolong
 * 创建时间：Created by  on 2020/6/23.
 *
 * 可关注公众号：我的大前端生涯   获取最新技术分享
 * 可关注网易云课堂：https://study.163.com/instructor/1021406098.htm
 * 可关注博客：https://blog.csdn.net/zl18603543572
 */

///lib/demo/search_textfield_bar.dart
///圆角搜索框
class SearchTextFieldBar extends StatefulWidget {
  ///搜索框上显示的文案
  String hint;
  ///hero过渡动画的tag
  String heroTag;
  ///搜索框的圆角
  double defaultBorderRadius;
  EdgeInsets margin;
  EdgeInsets padding;

  ///点击的回调事件
  Function clickCallBack;

  ///如果考虑不需要水波纹效果那么就可以设置颜色为透明色
  Color splashColor;
  ///文本输入框焦点控制
  FocusNode focusNode;
  ///文本输入框
  TextEditingController controller;
  ///点击键盘上的回车键的回调
  Function(String text) onSubmitted;
  ///输入文本的长度限制
  int inputKeyWordsLength;
  ///输入框文字大小
  double fontSize;

  SearchTextFieldBar(
      {this.hint = "搜索",
      this.defaultBorderRadius = 20.0,
      this.margin = const EdgeInsets.only(top: 10.0, bottom: 10.0),
      this.padding = const EdgeInsets.only(left: 12),
      this.heroTag,
      this.clickCallBack,
      this.splashColor = const Color(0xFFD6D6D6),
      this.focusNode,
      this.controller,
      this.onSubmitted,
      this.inputKeyWordsLength=20,
      this.fontSize = 14});

  @override
  State<StatefulWidget> createState() {
    return SearchTextFieldBarState();
  }
}

class SearchTextFieldBarState extends State<SearchTextFieldBar> {
  ///为true 时显示清除选项
  bool showClear = false;

  @override
  void initState() {
    super.initState();
    if (widget.focusNode == null) {
      widget.focusNode = new FocusNode();
    }
    if (widget.controller == null) {
      widget.controller = new TextEditingController();
    }
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      ///外边距
      margin: widget.margin,
      child: buildContainer(context),
    );
  }

  ///构建搜索框的显示区域[Container]
  Container buildContainer(BuildContext context) {
    return Container(
      height: widget.defaultBorderRadius * 2,

      ///获取当前StatelessWidget的宽度
      width: MediaQuery.of(context).size.width,

      ///对齐方式
      alignment: Alignment(-1, 0),

      ///内边距
      padding: widget.padding,

      ///圆角边框
      decoration: BoxDecoration(
        color: Colors.white,
        borderRadius: BorderRadius.circular(widget.defaultBorderRadius),
      ),
      child: buildRow(),
    );
  }

  ///构建搜索图标与显示文本
  Row buildRow() {
    return Row(
      ///居左
      mainAxisAlignment: MainAxisAlignment.start,
      crossAxisAlignment: CrossAxisAlignment.center,
      children: <Widget>[
        ///搜索图标
        Image.asset(
          "assets/images/2.0x/search_icon.png",
          width: 18.0,
          height: 18.0,
        ),
        SizedBox(
          width: 8.0,
        ),
        Expanded(
          flex: 1,
          child: Container(
            alignment: Alignment(-1, 0),
            child: TextField(
              ///设置键盘的类型
              keyboardType: TextInputType.text,
              ///键盘回车键的样式为搜索
              textInputAction: TextInputAction.search,
              ///只有苹果手机上有效果
              keyboardAppearance:Brightness.dark,
              controller: widget.controller,
              ///最大行数
              maxLines: 1,
              ///输入文本格式过滤
              inputFormatters: [
                ///输入的内容长度限制
                LengthLimitingTextInputFormatter(widget.inputKeyWordsLength),
              ],
              ///当输入文本时实时回调
              onChanged: (text) {
                ///多层判断 优化刷新
                ///只有当有改变时再刷新
                if (text.length > 0) {
                  if (!showClear) {
                    showClear = true;
                    setState(() {});
                  }
                } else {
                  if (showClear) {
                    showClear = false;
                    setState(() {});
                  }
                }
              },
              ///点击键盘上的回车键
              ///或者是搜索按钮的回调
              onSubmitted: (text) {
                if(widget.onSubmitted!=null){
                  widget.onSubmitted(text);
                }
              },
              ///输入框不自动获取焦点
              autofocus: false,
              focusNode: widget.focusNode,
              style: TextStyle(
                  fontSize: widget.fontSize,
                  color: Colors.black,
                  fontWeight: FontWeight.w300),
              ///输入框的边框装饰
              decoration: InputDecoration(
                  //重要 用于编辑框对齐
                  isDense: true,
                  ///清除文本内边跑
                  contentPadding: EdgeInsets.zero,
                  ///不设置边框
                  border: InputBorder.none,
                  ///设置提示文本
                  hintText: widget.hint,
                  ///设置提示文本的样式
                  hintStyle: TextStyle(
                    fontSize: widget.fontSize,
                    color: Color(0xff999999),
                  )),
            ),
          ),
        ),
        buildClearButton(),
      ],
    );
  }
  ///清除按键
  ///当文本框有内容输入时显示清除按钮
  buildClearButton() {
    if (showClear) {
      return IconButton(
        icon: Icon(
          Icons.clear,
          size: 24.0,
          color: Color(0xffacacac),
        ),
        onPressed: () {
          widget.controller.clear();
        },
      );
    } else {
      return Container();
    }
  }
}


lib/src/search_static_bar.dart
import 'dart:ui';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

/**
 * 创建人： Created by zhaolong
 * 创建时间：Created by  on 2020/6/23.
 *
 * 可关注公众号：我的大前端生涯   获取最新技术分享
 * 可关注网易云课堂：https://study.163.com/instructor/1021406098.htm
 * 可关注博客：https://blog.csdn.net/zl18603543572
 */

///lib/demo/search_static_bar.dart
///圆角搜索框显示
class SearchStaticBar extends StatelessWidget {
  ///搜索框上显示的文案
  String hint;
  ///hero过渡动画的tag
  String heroTag;
  ///搜索框的圆角
  double defaultBorderRadius;
  EdgeInsets margin;
  EdgeInsets padding;
  ///点击的回调事件
  Function clickCallBack;
  ///如果考虑不需要水波纹效果那么就可以设置颜色为透明色
  Color splashColor;

  SearchStaticBar(
      {this.hint = "搜索",
      this.defaultBorderRadius = 20.0,
      this.margin = const EdgeInsets.only(top: 10.0, bottom: 10.0),
      this.padding = const EdgeInsets.only(top: 8.0, bottom: 8.0),
      this.heroTag,
      this.clickCallBack,
      this.splashColor = const Color(0xFFD6D6D6)});
  @override
  Widget build(BuildContext context) {
    return Container(
      ///外边距
      margin: margin,
      child: Material(
        color: Colors.transparent,
        child: Ink(
            ///圆角边框
            decoration: BoxDecoration(
              color: Colors.white,
              borderRadius: BorderRadius.circular(defaultBorderRadius),
            ),
            child: InkWell(
              onTap: () {
                if (clickCallBack != null) {
                  clickCallBack();
                }
              },
              splashColor: splashColor,
              ///点击事件的圆角
              ///表现为水波纹的圆角
              borderRadius: BorderRadius.circular(defaultBorderRadius),

              ///页面过渡动画
              child: buildHeroWidget(context),
            )),
      ),
    );
  }


  ///构建[Hero]过渡动画
  Widget buildHeroWidget(BuildContext context) {
    if (heroTag == null) {
      return buildContainer(context);
    } else {
      return Hero(
        tag: heroTag,
        child: buildContainer(context),
      );
    }
  }
  ///构建搜索框的显示区域[Container]
  Container buildContainer(BuildContext context) {
    return Container(
      ///获取当前StatelessWidget的宽度
      width: MediaQuery.of(context).size.width,
      ///对齐方式
      alignment: Alignment(-1, 0),
      ///内边距
      padding: padding,
      child: buildRow(),
    );
  }
  ///构建搜索图标与显示文本
  Row buildRow() {
    return Row(
      ///居中
      mainAxisAlignment: MainAxisAlignment.center,
      children: <Widget>[
        ///搜索图标
        Image.asset(
          "assets/images/2.0x/search_icon.png",
          width: 14.0,
          height: 14.0,
        ),
        SizedBox(width: 8.0,),
        Text(
          hint,
          style: TextStyle(color: Color(0xff999999), fontSize: 16.0),
        ),
      ],
    );
  }
}

class SearchBar extends StatefulWidget implements PreferredSizeWidget {
  final ValueChanged<String> onChanged;
  final ValueChanged<String> onSearch;
  final double horizontalPadding;
  final bool home;
  final String hint;
  final bool backIcon;
  final bool isShowNavTop;

  ///是否显示最左侧的搜索字样
  final bool isShowSearchButton;

  const SearchBar({
    Key key,
    this.home = false,
    this.isShowNavTop = false,
    this.onChanged,
    this.onSearch,
    this.horizontalPadding: 32.0,
    this.hint = "搜索",
    this.backIcon = true,
    this.isShowSearchButton = true,
  }) : super(key: key);

  @override
  _SearchBarState createState() => _SearchBarState();

  @override
  // TODO: implement preferredSize
  Size get preferredSize => Size.fromHeight(48.0);
}

class _SearchBarState extends State<SearchBar> {
  bool showClear = false;
  bool showHint = false;
  FocusNode _focusNode = new FocusNode();
  final TextEditingController controller = TextEditingController();

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return _goSearch();
  }

  _goSearch() {
    return GestureDetector(
      onTap: () {},
      child: widget.home ? _goHomeSearch() : _goNormalSearch(),
    );
  }

  _goHomeSearch() {
    return Container(
      color: Color(0xfff2f2f2),
      padding: EdgeInsets.symmetric(horizontal: widget.horizontalPadding),
      child: _inputBox(),
    );
  }

  _goNormalSearch() {
    return Container(
      height: 58.0,
      color: Color(0xfff2f2f2),
      margin: EdgeInsets.only(
          top: widget.isShowNavTop
              ? MediaQueryData
              .fromWindow(window)
              .padding
              .top
              : 0.0),
      padding: EdgeInsets.only(right: 16.0),
      child: Row(
        crossAxisAlignment: CrossAxisAlignment.center,
        children: <Widget>[
          _wrapGesture(
              Container(
                padding: EdgeInsets.fromLTRB(10.0, 0.0, 10.0, 6.0),
                child: widget.backIcon
                    ? Icon(
                  Icons.arrow_back_ios,
                  color: Color(0xff1d1d1d),
                  size: 20.0,
                )
                    : Container(),
              ),
              leftButtonClick),
          Expanded(flex: 1, child: _inputBox()),
        ],
      ),
    );
  }

  _wrapGesture(Widget widget, void Function() callback) {
    return InkWell(
      onTap: callback,
      child: widget,
    );
  }

  _inputBox() {
    Color inputBoxBackgroundColor;
    inputBoxBackgroundColor = Colors.white;
    return Container(
      margin: EdgeInsets.only(top: 10.0, bottom: 10.0),
      alignment: Alignment(-1, 0),
      padding: EdgeInsets.fromLTRB(10.0, 0.0, 10.0, 0.0),
      decoration: BoxDecoration(
        color: inputBoxBackgroundColor,
        borderRadius: BorderRadius.circular(20.0),
      ),
      child: Stack(
        alignment: Alignment.centerLeft,
        children: <Widget>[
          Row(
            crossAxisAlignment: CrossAxisAlignment.center,
            children: <Widget>[
              widget.home
                  ? Container()
                  : Container(
                alignment: Alignment(-1, 0),
                padding: EdgeInsets.only(
                  left: 4.0,
                ),
                child: Image.asset(
                  "assets/images/2.0x/search_icon.png",
                  width: 15.0,
                  height: 15.0,
                ),
              ),
              SizedBox(
                width: 10,
              ),
              Expanded(
                flex: 1,
                child: widget.home
                    ? Container()
                    : Container(
                    alignment: Alignment(-1, 0),
                    child: TextField(
                      controller: controller,
                      onChanged: (text) => _onChange(text),
                      onSubmitted: (text) => _onSearch(text),
                      autofocus: false,
                      focusNode: _focusNode,
                      enabled: !widget.home,
                      style: TextStyle(
                          fontSize: 15.0,
                          color: Colors.black,
                          fontWeight: FontWeight.w300),
                      decoration: InputDecoration(
                          border: InputBorder.none,
                          contentPadding: EdgeInsets.only(bottom: 12),
                          hintText: widget.hint,
                          hintStyle: TextStyle(
                            fontSize: 15.0,
                            color: Color(0xff999999),
                          )),
                    )),
              ),
              showClear
                  ? _wrapGesture(
                  Icon(
                    Icons.clear,
                    size: 24.0,
                    color: Color(0xffacacac),
                  ), () {
                controller.clear();
                _onChange("");
                setState(() {});
              })
                  : Container(),
            ],
          ),
          showClear
              ? Container()
              : widget.home
              ? Container(
            padding: EdgeInsets.only(bottom: 0.0),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                Image.asset(
                  "assets/images/2.0x/search_icon.png",
                  width: 14.0,
                  height: 14.0,
                ),
                SizedBox(
                  width: 8.0,
                ),
                Text(
                  widget.hint,
                  style: TextStyle(
                      color: Color(0xff999999), fontSize: 16.0),
                ),
              ],
            ),
          )
              : Container(),
          buildSearchButton(),
        ],
      ),
    );
  }

  buildSearchButton() {
    return (widget.isShowSearchButton && showClear)
        ? Positioned(
      right: 30,
      child: InkWell(
        onTap: () {
          _onSearch(controller.text);
        },
        child: Container(
          alignment: Alignment.centerRight,
          width: 40,
          child: Text(
            "搜索",
            style: TextStyle(fontSize: 12, color: Colors.blue),
          ),
        ),
      ),
    )
        : Container();
  }

  _onChange(String text) {
    if (text.length > 0) {
      showClear = true;
    } else {
      showClear = false;
    }

    if (widget.onChanged != null) {
      widget.onChanged(text);
    }
    setState(() {});
  }

  _onSearch(String text) {
    ///注销焦点
    _focusNode.unfocus();
    if (widget.onSearch != null) {
      widget.onSearch(text);
    }
  }

  void leftButtonClick() {
    Navigator.pop(context);
  }
}


